ibtk/src/utilities/SnapshotCache.cpp

```diff
@@ -0,0 +1,251 @@
+                                         const double time,
+                                         Pointer<PatchHierarchy<NDIM> > current_hierarchy,
+                                         const std::string& snapshot_refine_type,
+                                         const double tol)
+{
+    // Make sure we are on a stored snapshot.
+    auto it =
+        std::find_if(d_idx_time_map.begin(),
+                     d_idx_time_map.end(),
+                     [time, tol](const std::pair<int, double>& t) -> bool { return std::abs(t.second - time) < tol; });
+    if (it == d_idx_time_map.end())
+        TBOX_ERROR("Snapshot at time: " << time << " with tolerance " << tol << " does not exist!\n");
+
+    // We have our snapshot index. Now copy the data to u_idx.
+    const std::pair<int, double>& snapshot_idx_t_pair = *it;
+    const int snapshot_idx = snapshot_idx_t_pair.first;
+    const Pointer<PatchHierarchy<NDIM> >& snapshot_hierarchy = d_snapshot_hierarchies.at(snapshot_idx);
+    const Pointer<VariableType>& snapshot_var = d_snapshot_vars.at(snapshot_idx);
+    // Make sure the current hierarchy and snapshot hierarchy have the same number of levels
+    TBOX_ASSERT(current_hierarchy->getFinestLevelNumber() == snapshot_hierarchy->getFinestLevelNumber());
+    // Now transfer data from the snapshot to the current hierarchy
+    int coarsest_ln = 0;
+    int finest_ln = snapshot_hierarchy->getFinestLevelNumber();
+    for (int ln = coarsest_ln; ln <= finest_ln; ++ln)
+    {
+        Pointer<PatchLevel<NDIM> > cur_level = current_hierarchy->getPatchLevel(ln);
+        Pointer<PatchLevel<NDIM> > snp_level = snapshot_hierarchy->getPatchLevel(ln);
+
+        // RefineSchedule requires that patch data be stored at the same points.
+        // We reset the snapshot patch data time to that of the current time.
+        double allocated_time =
+            cur_level->getPatch(*PatchLevel<NDIM>::Iterator(cur_level))->getPatchData(u_idx)->getTime();
+        // Note we need scratch allocated on the current hierarchy. Use the snapshot for that.
+        cur_level->allocatePatchData(snapshot_idx, allocated_time);
+        for (PatchLevel<NDIM>::Iterator p(snp_level); p; p++)
+        {
+            Pointer<Patch<NDIM> > patch = snp_level->getPatch(p());
+            Pointer<PatchData<NDIM> > data = patch->getPatchData(snapshot_idx);
+            data->setTime(allocated_time);
+        }
+
+        // Now copy the data.
+        Pointer<RefineAlgorithm<NDIM> > refine_alg = new RefineAlgorithm<NDIM>();
+        Pointer<CartesianGridGeometry<NDIM> > grid_geom = current_hierarchy->getGridGeometry();
+        Pointer<RefineOperator<NDIM> > refine_op = grid_geom->lookupRefineOperator(snapshot_var, snapshot_refine_type);
+        refine_alg->registerRefine(u_idx, snapshot_idx, snapshot_idx, refine_op);
+        Pointer<RefineSchedule<NDIM> > schedule =
+            refine_alg->createSchedule(cur_level, snp_level, ln - 1, current_hierarchy);
```

Comment:
If these get re-used, it might be worth saving them.

Reply:
The though has crossed my mind. If we're pulling a lot of snapshots out of this cache, we'll potentially be doing a lot of transfers which could be expensive. I'm not sure of a good way to save the schedule's. We would have to determine if `current_hierarchy` hasn't changed it's patch configuration.

I'd rather tackle this problem later if it becomes an issue.